### PR TITLE
let CliGitAPIImpl rely on FilePath+Launcher remote API

### PR DESCRIPTION
**Pull request by ndeloof**
This (long) pull-request is about making CLIGitAPIImpl rely on FilePath and Launcher and have logic run on master JVM (as expected by the remoting abstraction). As JGitAPIImpl relies on JGit - so on java.io.File - it has to run fully remotely, so had to split logic between implementations.

Most changes are about moving from File to FilePath and adding IOException to API methods.

Also removed getRepository that was deprecated since 1.1 as not being remotable.

Final goal is to let git-client rely on configured Launcher, which hold the computer environment or customized behavior from caller (using DecoratedLauncher)
see https://issues.jenkins-ci.org/browse/JENKINS-30600, also applies to git ran inside Pipeline as computer's NodeProperties are exposed to git by Launcher.

### CODE REVIEW

**Comment by ndeloof** [comment]
Not sure about CliGitAPIImplTest.test_checkout_interrupted_with_existing_lock failure on CI. Works on my machine (I hate to say that), maybe a race condition ?

**Comment by ndeloof** [comment]
@reviewbybees

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 1]
```diff
@@ -1872,29 +1901,8 @@ public void execute() throws GitException, InterruptedException {
...
-            if (branches.size() == 1) {
-                listener.getLogger().println("Seen 1 remote branch");
-            } else {
-                listener.getLogger().println(MessageFormat.format("Seen {0} remote branches", branches.size()));
-            }
-
-            return branches;
-        }
+    public Set<Branch> getRemoteBranches() throws GitException, InterruptedException, IOException {
+        return parseBranches(launchCommand("branch", "-r", "-v", "--no-abbrev"));
```
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: wonder why the initial code was relying on JGit within the CLIGit implementation. Would prefer not to mix such changes in this PR, but had no choice here.

**Comment by ndeloof** [comment]
@MarkEWaite test failure fixed, open for review

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 2]
```diff
@@ -1486,15 +1512,11 @@ private String launchCommandWithCredentials(ArgumentListBuilder args, File workD
...
-            List<String> privateKeys = sshUser.getPrivateKeys();
-            for (String s : privateKeys) {
-                w.println(s);
-            }
-        }
-        new FilePath(key).chmod(0400);
+    private FilePath createSshKeyFile(SSHUserPrivateKey sshUser) throws IOException, InterruptedException {
+        FilePath key = workspace.createTempFile("ssh", "key");
+        List<String> privateKeys = sshUser.getPrivateKeys();
+        key.write(StringUtils.join(privateKeys, "\n"), null);
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: smells like a bug to me the OS may be windows which does not use '\n' but '\r\n'

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 3]
```diff
@@ -1510,51 +1532,42 @@ private String quoteUnixCredentials(String str) {
...
-            w.println("@if (%arg:~0,8%)==(Username) echo " + quoteWindowsCredentials(creds.getUsername()));
-            w.println("@if (%arg:~0,8%)==(Password) echo " + quoteWindowsCredentials(Secret.toString(creds.getPassword())));
-        }
-        askpass.setExecutable(true);
+    private FilePath createWindowsStandardAskpass(StandardUsernamePasswordCredentials creds) throws IOException, InterruptedException {
+        FilePath askpass = workspace.createTempFile("pass", ".bat");
+        askpass.write("@set arg=%~1\n"
+                    + "@if (%arg:~0,8%)==(Username) echo " + quoteWindowsCredentials(creds.getUsername()) + "\n"
+                    + "@if (%arg:~0,8%)==(Password) echo " + quoteWindowsCredentials(Secret.toString(creds.getPassword())), null);
+        askpass.chmod(0700);
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: what do you expect this to do on windows :-)

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 4]
```diff
@@ -1510,51 +1532,42 @@ private String quoteUnixCredentials(String str) {
...
-        File ssh = File.createTempFile("pass", ".bat");
-        try (PrintWriter w = new PrintWriter(ssh)) {
-            w.println("echo \"" + quoteWindowsCredentials(Secret.toString(sshUser.getPassphrase())) + "\"");
-            w.flush();
-        }
-        ssh.setExecutable(true);
+    private FilePath createWindowsSshAskpass(SSHUserPrivateKey sshUser) throws IOException, InterruptedException {
+        FilePath ssh = workspace.createTempFile("pass", ".bat");
+        ssh.write("echo \"" + Secret.toString(sshUser.getPassphrase()) + "\"", null);
+        ssh.chmod(0700);
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: what do you expect this to do on windows :-)

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 2 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: right, but git on windows relies on openssl and can read unix EOF without issue

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 4 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: FilePath has no equivalent for setExecutable I know, I'm open to any guidance here

Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: I'm not sure that we want the unencrypted ssh passphrase written into the workspace.  Isn't that even more widely available than the system temporary directory (through the "Workspace" browsing facilities in Jenkins).

I'm considering switching from `echo` to `type` and `cat` so that the shell escaping rules of echo don't have to be applied to the passphrase.  Refer to https://github.com/MarkEWaite/git-client-plugin/pull/2/commits/22ae93a9f5d1088fb353995e69bda2cb0927c839 for the beginnings of that code.

I'd rather have this change (modify the location of the passphrase) be considered in the context of #207, since it is also touching the same area, but touching it in a way that fixes a Windows problem.

Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: The chmod (and setExecutable) are not needed on Windows.  Files that end with the `bat` suffix are executable.

Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: should be in `workspace@tmp`

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 5]
```diff
@@ -1828,14 +1838,15 @@ public void execute() throws GitException, InterruptedException {
      */
     private Set<Branch> parseBranches(String fos) {
         Set<Branch> branches = new HashSet<>();
+        Set<String> links = new HashSet<String>();
         BufferedReader rdr = new BufferedReader(new StringReader(fos));
         String line;
         try {
             while ((line = rdr.readLine()) != null) {
-                if (line.length() < 44 || !line.contains(" ")) {
+                if (line.length() < 5 || !line.contains(" ")) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: Seems a bit unrelated and a big change?

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 6]
```diff
@@ -1846,10 +1857,24 @@ public void execute() throws GitException, InterruptedException {
...
                 // Fields are separated by one or more spaces
                 String[] branchVerboseOutput = line.substring(2).split(" +", 3);
+                if (branchVerboseOutput[1].equals("->")) {
+                    links.add(line);
+                }
                 if (branchVerboseOutput[1].length() == 40) {
                     branches.add(new Branch(branchVerboseOutput[0], ObjectId.fromString(branchVerboseOutput[1])));
                 }
             }
+            for (String link : links) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: seems unrelated to the change in question?

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 4 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: Moving to workspace@tmp is out of scope for this PR. 
@MarkEWaite tmp file is deleted on finally block, isn't this safe enough ? If you don't think so, can rely on FilePath#createTextTempFile which uses java.io.tmpdir

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 6 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: indeed. I had to re-implement getRemoteBranches in plain git cli as the legacy code was relying on JGIT for this CLI implementation (sic), and JGit doesn't behave the same as `git branch -r`.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 5 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: JGIT do report HEADs as plain branches, while git CLI do use `->` and reference to another branch

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 4 above)
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: If (as @jtnord says) workspace.createTempFile() creates the file in a directory which is not visible from the workspace link in the Jenkins job, then that makes your change much better than the previous implementation.  The previous implementation wrote a plaintext secret to java.io.tmpdir, while your change writes a plaintext secret to a file into that directory hidden inside the Jenkins workspace.

Both the old and new implementation remove the plaintext secret file in the finally block.

Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: applied 2c321ce8334176270983d3b59cda8fb148f8a314 to rely on createTextTempFile so changes here just act as the initial implementation. Topic about moving this tmp file to another (better?) place is unrelated.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 7]
```diff
@@ -1510,51 +1532,41 @@ private String quoteUnixCredentials(String str) {
...
-        try (PrintWriter w = new PrintWriter(askpass)) {
-            w.println("@set arg=%~1");
-            w.println("@if (%arg:~0,8%)==(Username) echo " + quoteWindowsCredentials(creds.getUsername()));
-            w.println("@if (%arg:~0,8%)==(Password) echo " + quoteWindowsCredentials(Secret.toString(creds.getPassword())));
-        }
-        askpass.setExecutable(true);
+    private FilePath createWindowsStandardAskpass(StandardUsernamePasswordCredentials creds) throws IOException, InterruptedException {
+        FilePath askpass = workspace.createTextTempFile("pass", ".bat",
+                  "@set arg=%~1\r\n"
+                + "@if (%arg:~0,8%)==(Username) echo " + quoteWindowsCredentials(creds.getUsername()) + "\n\n"
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: `\n\n`?

Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: good catch

**Comment by ndeloof** [comment]
rebased

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 8]
```diff
@@ -1182,8 +1208,8 @@ public boolean isBareRepository(String GIT_DIR) throws GitException, Interrupted
         return !"false".equals(line);
     }
 
-    public boolean isShallowRepository() {
-        return new File(workspace, pathJoin(".git", "shallow")).exists();
+    public boolean isShallowRepository() throws IOException, InterruptedException {
+        return workspace.child(".git/shallow").exists();
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: NIT:  `child(".git").child("shallow").exists()`.

Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: FilePath do internally handle path separator normalization
(https://github.com/jenkinsci/jenkins/blob/fdabb525034e587bab2ec56eb7fc4c2431908232/core/src/main/java/hudson/FilePath.java#L265-L265)

File: src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java [hunk 9]
```diff
@@ -850,14 +873,23 @@ public String getRemoteUrl(String name) throws GitException {
...
         try {
             return new RepositoryBuilder().setWorkTree(workspace).build();
         } catch (IOException e) {
             throw new GitException(e);
         }
     }
 
+    public <T> T withRepository(RepositoryCallback<T> callable) throws IOException, InterruptedException {
+        Repository repo = getRepository();
+        try {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java]: if repository is `AutoCloseable` then try-with-resources is better here - if both the callable and the close fails the initial error would be swallowed with the code as is :-(

File: src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java [hunk 10]
```diff
@@ -1813,6 +1845,8 @@ else if (refspec != null)
             revListCommand.execute();
         } catch (InterruptedException e) {
             throw new GitException(e);
+        } catch (IOException e) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java]: multi-catch

File: src/main/java/org/jenkinsci/plugins/gitclient/LegacyCompatibleGitAPIImpl.java [hunk 11]
```diff
@@ -144,51 +120,58 @@ public void push(String remoteName, String refspec) throws GitException, Interru
...
-            final List<Tag> ret = new ArrayList<>();
-
-            for (final Map.Entry<String, Ref> tag : db.getTags().entrySet()) {
-                Ref value = tag.getValue();
-                if (value != null) {
-                    final ObjectId tagId = value.getObjectId();
-                    if (commit != null && commit.equals(tagId))
-                        ret.add(new Tag(tag.getKey(), tagId));
+    public List<Tag> getTagsOnCommit(final String revName) throws GitException, IOException, InterruptedException {
+        return withRepository(new RepositoryCallback<List<Tag>>() {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/LegacyCompatibleGitAPIImpl.java]: Seems like this needs to be a static inner clsas otherwise the outer Object will be serialized.  (not sure if this gets sent to run on a slave - but it smells like it might...)

Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/LegacyCompatibleGitAPIImpl.java]: I didn't tried to fix those Deprecated methods, would prefer to remove them all but this would make this PR even harder to review

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 12]
```diff
@@ -1496,15 +1522,11 @@ private String launchCommandWithCredentials(ArgumentListBuilder args, File workD
...
-        key = File.createTempFile("ssh", "key");
-        try (PrintWriter w = new PrintWriter(key)) {
-            List<String> privateKeys = sshUser.getPrivateKeys();
-            for (String s : privateKeys) {
-                w.println(s);
-            }
-        }
-        new FilePath(key).chmod(0400);
+    private FilePath createSshKeyFile(SSHUserPrivateKey sshUser) throws IOException, InterruptedException {
+        FilePath key = workspace.createTempFile("ssh", "key");
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: this smells insecure.  use [`WorkspaceList.tmpDir(workspace)`](https://github.com/jenkinsci/jenkins/commit/307bfc17385c14aefcb623ccc3e7144e4f01e2a3)

If you can not bump a version to that write a helper that does the same thing.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 13]
```diff
@@ -1520,51 +1542,41 @@ private String quoteUnixCredentials(String str) {
...
 
-    private File createWindowsSshAskpass(SSHUserPrivateKey sshUser) throws IOException {
-        File ssh = File.createTempFile("pass", ".bat");
-        try (PrintWriter w = new PrintWriter(ssh)) {
-            w.println("echo \"" + quoteWindowsCredentials(Secret.toString(sshUser.getPassphrase())) + "\"");
-            w.flush();
-        }
-        ssh.setExecutable(true);
+    private FilePath createWindowsSshAskpass(SSHUserPrivateKey sshUser) throws IOException, InterruptedException {
+        FilePath ssh = workspace.createTextTempFile("pass", ".bat",
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: this smells insecure.  use [`WorkspaceList.tmpDir(workspace)`](https://github.com/jenkinsci/jenkins/commit/307bfc17385c14aefcb623ccc3e7144e4f01e2a3)

If you can not bump a version to that write a helper that does the same thing.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 14]
```diff
@@ -1520,51 +1542,41 @@ private String quoteUnixCredentials(String str) {
...
 
-    private File createUnixSshAskpass(SSHUserPrivateKey sshUser) throws IOException {
-        File ssh = File.createTempFile("pass", ".sh");
-        try (PrintWriter w = new PrintWriter(ssh)) {
-            w.println("#!/bin/sh");
-            w.println("echo '" + quoteUnixCredentials(Secret.toString(sshUser.getPassphrase())) + "'");
-        }
-        ssh.setExecutable(true);
+    private FilePath createUnixSshAskpass(SSHUserPrivateKey sshUser) throws IOException, InterruptedException {
+        FilePath ssh = workspace.createTextTempFile("pass", ".sh",
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: this smells insecure.  use [`WorkspaceList.tmpDir(workspace)`](https://github.com/jenkinsci/jenkins/commit/307bfc17385c14aefcb623ccc3e7144e4f01e2a3)

If you can not bump a version to that write a helper that does the same thing.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 15]
```diff
@@ -1520,51 +1542,41 @@ private String quoteUnixCredentials(String str) {
...
-    private File createWindowsStandardAskpass(StandardUsernamePasswordCredentials creds) throws IOException {
-        File askpass = File.createTempFile("pass", ".bat");
-        try (PrintWriter w = new PrintWriter(askpass)) {
-            w.println("@set arg=%~1");
-            w.println("@if (%arg:~0,8%)==(Username) echo " + quoteWindowsCredentials(creds.getUsername()));
-            w.println("@if (%arg:~0,8%)==(Password) echo " + quoteWindowsCredentials(Secret.toString(creds.getPassword())));
-        }
-        askpass.setExecutable(true);
+    private FilePath createWindowsStandardAskpass(StandardUsernamePasswordCredentials creds) throws IOException, InterruptedException {
+        FilePath askpass = workspace.createTextTempFile("pass", ".bat",
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: this smells insecure.  use [`WorkspaceList.tmpDir(workspace)`](https://github.com/jenkinsci/jenkins/commit/307bfc17385c14aefcb623ccc3e7144e4f01e2a3)

If you can not bump a version to that write a helper that does the same thing.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 16]
```diff
@@ -1520,51 +1542,41 @@ private String quoteUnixCredentials(String str) {
...
-        try (PrintWriter w = new PrintWriter(askpass)) {
-            w.println("#!/bin/sh");
-            w.println("case \"$1\" in");
-            w.println("Username*) echo '" + quoteUnixCredentials(creds.getUsername()) + "' ;;");
-            w.println("Password*) echo '" + quoteUnixCredentials(Secret.toString(creds.getPassword())) + "' ;;");
-            w.println("esac");
-        }
-        askpass.setExecutable(true);
+    private FilePath createUnixStandardAskpass(StandardUsernamePasswordCredentials creds) throws IOException, InterruptedException {
+        FilePath askpass = workspace.createTextTempFile("pass", ".sh",
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: this smells insecure.  use [`WorkspaceList.tmpDir(workspace)`](https://github.com/jenkinsci/jenkins/commit/307bfc17385c14aefcb623ccc3e7144e4f01e2a3)

If you can not bump a version to that write a helper that does the same thing.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 17]
```diff
@@ -1681,44 +1695,39 @@ private File getSSHExeFromGitExeParentDir(String userGitExe) {
         throw new RuntimeException("ssh executable not found. The git plugin only supports official git client http://git-scm.com/download/win");
     }
 
-    private File createWindowsGitSSH(File key, String user) throws IOException {
-        File ssh = File.createTempFile("ssh", ".bat");
+    private FilePath createWindowsGitSSH(FilePath key, String user) throws IOException, InterruptedException {
+        FilePath ssh = workspace.createTempFile("ssh", ".bat");
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: not insecure - but no need to clutter the workspace.
use [`WorkspaceList.tmpDir(workspace)`](https://github.com/jenkinsci/jenkins/commit/307bfc17385c14aefcb623ccc3e7144e4f01e2a3)

If you can not bump a version to that write a helper that does the same thing.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 13 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: My goal here is to make git-client remote-compliant, and doing this I keep the existing implementation decision. Relying on workspace@tmp is another topic.

**Comment by jtnord** [comment]
:bug: there is a lot that smells insecure to me - and a lot of places where the workspace is littered with temporary files without due need.

Use [`WorkspaceList.tmpDir(workspace)`](http://javadoc.jenkins-ci.org/hudson/slaves/WorkspaceList.html#tempDir-hudson.FilePath-) if you can use a parent >= 1.652  

If you can not bump a version to that write a helper that does the same thing.

**Comment by ndeloof** [comment]
@jtnord removed use of createTempFile so temp files are created under /tmp (as it was previously)

use of workspace@tmp is another (arguable) topic I don't plan to address here

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 13 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: no if this introduces a security vulnerability (by being able to retrive credentials/keys via the workspace browser) the it is not another topic at all.

**Comment by jtnord** [comment]
@ndeloof check the javadoc for [`FilePath.createTempFile()`](http://javadoc.jenkins-ci.org/hudson/FilePath.html#createTextTempFile-java.lang.String-java.lang.String-java.lang.String-)

Creates a temporary file in **this** directory and set the contents to the given text (encoded in the platform default encoding)

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 12 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: ^^ still in the workspace...

**Comment by jtnord** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 13 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: So the use of `createTextTempFile(..., false)`, so java.io.tmpDir is being used for temporary files.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 18]
```diff
@@ -2317,19 +2314,15 @@ private void env(String name, String value) {
...
-    @NonNull
-    public Repository getRepository() throws GitException {
-        try {
-            return FileRepositoryBuilder.create(new File(workspace, Constants.DOT_GIT));
-        } catch (IOException e) {
-            throw new GitException("Failed to open Git repository " + workspace, e);
-        }
+    @Override
+    public <T> T withRepository(final RepositoryCallback<T> callable) throws IOException, InterruptedException {
+        return workspace.act(new MasterToSlaveFileCallable<T>() {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: stephenc: > I am concerned about the anonymous inner class here that may pull over the outer class reference. Does it make sense to switch to an explicit named static inner class so that we clearly identify the minimal set of class references to send over the remoting reference

File: src/main/java/org/jenkinsci/plugins/gitclient/Git.java [hunk 19]
```diff
@@ -116,23 +126,37 @@ public Git using(String exe) {
...
-        jenkins.MasterToSlaveFileCallable<GitClient> callable = new jenkins.MasterToSlaveFileCallable<GitClient>() {
-            public GitClient invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
-                if (listener == null) listener = TaskListener.NULL;
-                if (env == null) env = new EnvVars();
 
-                if (exe == null || JGitTool.MAGIC_EXENAME.equalsIgnoreCase(exe)) {
+        if (env == null) env = new EnvVars();
+        if (repository == null) {
+            // Assume local directory
+            if (launcher != null) {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: Note to self - the launcher "." directory is the directory from which java -jar slave.jar was invoked, not the slave home directory, and not the workspace directory.  Add a comment to make that clear

File: src/main/java/org/jenkinsci/plugins/gitclient/Git.java [hunk 20]
```diff
@@ -116,23 +126,37 @@ public Git using(String exe) {
...
     public GitClient getClient() throws IOException, InterruptedException {
-        jenkins.MasterToSlaveFileCallable<GitClient> callable = new jenkins.MasterToSlaveFileCallable<GitClient>() {
-            public GitClient invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
-                if (listener == null) listener = TaskListener.NULL;
-                if (env == null) env = new EnvVars();
 
-                if (exe == null || JGitTool.MAGIC_EXENAME.equalsIgnoreCase(exe)) {
+        if (env == null) env = new EnvVars();
+        if (repository == null) {
+            // Assume local directory
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: Note to self: move this comment to the `else` branch

File: src/main/java/org/jenkinsci/plugins/gitclient/Git.java [hunk 21]
```diff
@@ -116,23 +126,37 @@ public Git using(String exe) {
...
+            if (launcher != null) {
+                repository = new FilePath(launcher.getChannel(), ".");
+            } else {
+                repository = new FilePath(new File("."));
+            }
+        }
+        GitClient git;
+
+        if (exe == null || JGitTool.MAGIC_EXENAME.equalsIgnoreCase(exe)) {
+            FileCallable<GitClient> callable = new MasterToSlaveFileCallable<GitClient>() {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: stephenc: > fairly certain this should be a named static inner class not anonymous inner

File: src/main/java/org/jenkinsci/plugins/gitclient/Git.java [hunk 22]
```diff
@@ -116,23 +126,37 @@ public Git using(String exe) {
...
+            // Assume local directory
+            if (launcher != null) {
+                repository = new FilePath(launcher.getChannel(), ".");
+            } else {
+                repository = new FilePath(new File("."));
+            }
+        }
+        GitClient git;
+
+        if (exe == null || JGitTool.MAGIC_EXENAME.equalsIgnoreCase(exe)) {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: @ndeloof Does the exe == null condition here mean  that now the default git implementation on an agent is JGit, whether or not the job selected JGit as the implementation?

If so, that's bad.  Users want their job selected git implementation honored, even if it is the system default.

File: src/main/java/org/jenkinsci/plugins/gitclient/Git.java [hunk 23]
```diff
@@ -116,23 +126,37 @@ public Git using(String exe) {
...
+        } else {
+            // Ensure we return a backward compatible GitAPI, even API only claim to provide a GitClient
+            git = new GitAPI(exe, repository, listener, env, launcher);
+        }
+
         Jenkins jenkinsInstance = Jenkins.getInstance();
         if (jenkinsInstance != null && git != null)
             git.setProxy(jenkinsInstance.proxy);
         return git;
+
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: Remove this added blank line

File: src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java [hunk 24]
```diff
@@ -47,6 +47,7 @@
 
 import javax.annotation.Nullable;
 
+import jenkins.model.Jenkins;
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java]: stephenc: > I am concerned greatly that you are importing Jenkins as that is a class that *should* (never mind that other plugins such as Subversion are bold and do bad things, but we should be getting better not worse) not be loading on agents. Given that we explicitly construct this class on a remote agent I would expect to see sufficient classloading guard methods to ensure that the methods referencing `Jenkins` are never called on execution paths which will be invoked on the agent and are only for execution paths running on the `Jenkins` master JVM... that is almost impossible to do without specifying a core dependency of 1.653+ (see `JenkinsJVM` class added in 1.653) so I think this is a bad change

Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java]: @ndeloof 

stephenc: > I also believe given the usage of this class in [this method](https://github.com/jenkinsci/git-client-plugin/pull/212/files#diff-88af64f5dbd015aa6e8f00f355f34700R146) the intent is that the JGitAPIImpl is supposed to be instantiated on the agent and not on the master. As such I would expect to see a `writeReplace` method that replaces `this` with `Channel.export(GitAPI.class,this)` in those cases where `Channel.current() != null`... further as I suspect that the credentials conversion in this class is probably written to assume that it is being invoked on the master JVM it may be necessary to wrap the exported proxy object in a delegating `Serializable` hand rolled proxy that unencrypts the credentials before passing through to the remoting proxy so that the decryption happens on the master (which is the only place that has access to the secret key)

File: src/main/java/org/jenkinsci/plugins/gitclient/LegacyCompatibleGitAPIImpl.java (hunk 11 above)
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/LegacyCompatibleGitAPIImpl.java]: stephenc: > static inner class++

File: src/main/java/org/jenkinsci/plugins/gitclient/RemoteGitImpl.java [hunk 25]
```diff
@@ -140,9 +141,9 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
             throw new IllegalStateException("Unexpected invocation: "+method);
         }
 
-        public void execute() throws GitException, InterruptedException {
+        public void execute() throws GitException, InterruptedException, IOException {
             try {
-                channel.call(new jenkins.security.MasterToSlaveCallable<Void, GitException>() {
+                channel.call(new MasterToSlaveCallable<Void, GitException>() {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/RemoteGitImpl.java]: stephenc: > static inner class

File: src/main/java/org/jenkinsci/plugins/gitclient/Git.java (hunk 22 above)
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: @MarkEWaite I haven't changed the condition here

Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/gitclient/Git.java]: @MarkEWaite I haven't changed the condition

**Comment by ndeloof** [comment]
build on ci.jenkins.io fails due to missing `git config --global user.email`.

**Comment by ndeloof** [comment]
@MarkEWaite why did you closed this PR ?

**Comment by MarkEWaite** [comment]
It was closed by GitHub as a side effect of my created the 1.21.x branch and merging the 2.0.0-beta branch to the master branch, then deleting the 2.0.0-beta branch.  Since git client 2.0.0 has released, could you reopen this pull request against the master branch?  Evaluation can then continue against the master branch.
